### DISCUSSION TITLE
Photometric interpretation seems to return RGB incorrectly?

### DISCUSSION THREAD

**Discussion by miriam-groeneveld** [libvips/libvips#2341] [Q&A]
I have several files (svs, tiff) that have Photometric interpretation YCbCr, output from tiffinfo:

```
TIFF Directory at offset 0x377a3b4 (58172340)
  Image Width: 61843 Image Length: 26473
  Tile Width: 128 Tile Length: 128
  Resolution: 19980, 19980 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: JPEG
  Photometric Interpretation: YCbCr
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
  Reference Black/White:
     0:     0   255
     1:   128   255
     2:   128   255
  JPEG Tables: (574 bytes)
  ...
```

When I load the image with Image.new_from_file() the interpretation of the image becomes RGB. What is the reasoning behind this and is there a way I can find the original interpretation?

**Comment by jcupitt** [comment]
Hi @miriam-groeneveld,

libvips doesn't support YCbCr colourspace, so the TIFF (and JPEG) loaders convert to RGB. On save, it'll save as YCbCr if there are three uint8 bands, Q < 90 and JPEG compression is enabled. Things like `PHOTOMETRIC_LOGLUV` and `MINISWHITE` are handled in a similar way.

We could attach the original photometric interpretation as metadata I suppose. What's the use-case?

**Reply by jcupitt** [reply]
(to answer your question, no, sorry there is no current way to find this information. `tiffinfo` is probably your best bet)

**Reply by miriam-groeneveld** [reply]
Hi John, thanks for your reply. We convert a number of file formats to tiff (and then dzi), and would like to make sure that this tiff file has the same colour encoding as the original file. Not sure if this is possible without losing some of the data, as you say the loaders convert to RGB already. We would at least want to be able to save as YCbCr when the original was YCbCr, same for RGB  (and other colour encodings, I know the photometric interpretation in tiff has about 12 options).

**Reply by jcupitt** [reply]
I suppose these are jpeg-compressed TIFFs, is that right? libvips will write YCbCr by default, even if the source was RGB.

```
$ vips copy k2.png x.tif[compression=jpeg]
$ tiffinfo x.tif | grep Photo
  Photometric Interpretation: YCbCr
```

Same for DZI conversion.

Could you explain why you need to keep the photometric interpretation? I don't think YCbCr and RGB are meaningfully different.

**Reply by miriam-groeneveld** [reply]
Our platform allows people to create algorithms that have images as input, to keep things simple the platform converts these images to tiff/dzi. The algorithms expect a certain photometric interpretation (in most cases RGB), but we noticed in some cases the conversion creates tiff files with a different photometric interpretation than the original. This can cause the algorithm to fail. I noticed the argument rgbjpeg in the write_to_file function, which I would like to set to true only if the original was rgb.

**Reply by jcupitt** [reply]
I think that's probably a minor bug in your algorithms.

If a program needs RGB data from a TIFF, it has to set the colour mode. Add a line like this to your TIFF loader:

```
TIFFSetField( tiff, TIFFTAG_JPEGCOLORMODE, JPEGCOLORMODE_RGB );
```

It's a libtiff pseudo tag that controls the behaviour of the JPEG decoder.

**Reply by miriam-groeneveld** [reply]
Well, it's an open platform, we don't create the algorithms ourselves. It's just that we provide the conversion, and that changes the photometric interpretation.

**Reply by jcupitt** [reply]
If you save JPEG TIFFs with RGB photometric interpretation, you lose chrominance subsampling. For example: 

```
john@banana ~/pics $ vips copy marble.jpg x2.tif[compression=jpeg]
john@banana ~/pics $ convert marble.jpg -compress jpeg x1.tif
john@banana ~/pics $ ls -l x1.tif x2.tif
-rw-rw-r-- 1 john john 20059091 Jul 12 13:48 x1.tif
-rw-rw-r-- 1 john john  7229642 Jul 12 13:48 x2.tif
john@banana ~/pics $ 
```

The RGB photometric interpretation one is almost 3x larger for the same visual quality. 

Really there are two errors here -- people are uploading JPEG TIFFs with an incorrect photometric interpretation, and they are failing to request RGB data when they need RGB data.

That doesn't help you, since you clearly need to support these users. Perhaps the simplest solution would be for you to always save with Q of 90 or above? For example:

```
john@banana ~/pics $ vips copy k2.jpg x.tif[compression=jpeg,Q=90]
john@banana ~/pics $ tiffinfo x.tif
TIFF Directory at offset 0xfb55c (1029468)
  Image Width: 1450 Image Length: 2048
  Resolution: 72.009, 72.009 pixels/inch
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: JPEG
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Rows/Strip: 128
  Planar Configuration: single image plane
  JPEG Tables: (289 bytes)
```

Would that solve this problem? Your files will be much larger, of course. You can use the default Q 75 for the DZIs.

**Reply by miriam-groeneveld** [reply]
And the same applies for converting Aperio (svs) files I guess? Thank you for the explanation, the increase in file size will be something to consider.

**Reply by jcupitt** [reply]
Yes, just use a large Q value.
